### DISCUSSION TITLE
Finding entity with embeddables

### DISCUSSION THREAD

**Discussion by edorgeville** [mikro-orm/mikro-orm#3530] [Q&A]
Hi,
I am trying to query entities based on embeddables. Consider the following implementation for the least efficient online shop:
```typescript
console.log('create all elements')
const extraLargeBox = new entities.Packaging('box')
const pen = new entities.Product('Pen (1pcs)')
const inefficientPackage = new entities.Package(extraLargeBox, pen)

console.log('create an order and persist it')
const order = new entities.Order()
order.address = '10 Hello St.'
order.packages = [inefficientPackage]
await orm.em.persistAndFlush(order)

console.log('find the order from the package')
console.log(await orm.em.findOne(entities.Order, { packages: [inefficientPackage] }))
```
When running, this error shows up for the last line:
> ValidationError: Using operators inside embeddables is not allowed, move the operator above. (property: Order.packages, payload: { packages: { '$in': [ [Package] ] } })

I am unsure what "moving the operator above" means. Am I querying 
Here is the full repro:
https://github.com/edorgeville/mikro-orm-find-embeddables

Thanks in advance!

**Comment by B4nan** [comment]
`{ packages: [inefficientPackage] }` this query expands to `{ packages: { $in: [inefficientPackage] } }` which is use of operator inside nested embeddable properties - you cant do that.

Maybe try just `{ packages: inefficientPackage }`?

**Reply by edorgeville** [reply]
I just tried with `{ packages: inefficientPackage }`, it returns `null`. In this case it is a single package, but I would like to find a solution that covers multiple packages as well, in order to do:
```typescript
await orm.em.findOne(entities.Order, { packages: [package1, package2, package3] })
```

**Reply by B4nan** [reply]
Again, this is not supported, you cant query for array items, that array is always translated to `$in` query.

**Reply by B4nan** [reply]
I guess you could store only some identifiers in a top level property of the entity you could use for querying.

**Reply by edorgeville** [reply]
Gotcha. ~~Any obvious workarounds I might be missing? Otherwise I will look into modelling our data differently.~~
Edit: Just saw your answer. Thanks for your help!

**Reply by edorgeville** [reply]
Is it possible to disable query expansion so arrays don't get transformed into `$in`? Or should I use the native mongo driver directly?

**Reply by B4nan** [reply]
I dont think so, but we could probably find some easy way to do it, maybe via `expr` helper that would tag the value with some hidden symbol property.

**Reply by edorgeville** [reply]
In native MQL, the filter looks like this:
```javascript
{
  "packages": {
    $eq: [
    {
      "packaging": ObjectId("632cb6d90f37cc58b5f811fb"),
      "content": ObjectId("632cb6e50f37cc58b5f811fc")
    },
    {
      "packaging": ObjectId("632cb6e90f37cc58b5f811fd"),
      "content": ObjectId("632cb6fe0f37cc58b5f811fe")
    }
  ]
  }
}
```
I was hoping to be able to use ObjectIds directly:
```typescript
console.log('try using object ids')
const packages = [inefficientPackage]
const packagesUsingIds = packages.map((p) => {
  return {
    packaging: p.packaging._id,
    content: p.content._id
  }
})
console.log(await orm.em.findOne(entities.Order, { packages: { $eq: packagesUsingIds } }))
```
But the same error occurs because of the array expansion to `$in`

**Reply by edorgeville** [reply]
> I dont think so, but we could probably find some easy way to do it, maybe via `expr` helper that would tag the value with some hidden symbol property.

I am completely lost as to how to use the `expr` helper 😅

**Reply by edorgeville** [reply]
Here's a workaround I got for now using aggregations. 2 queries instead of 1, so not really ideal, but at least it works 🙃
```typescript
console.log('use aggregation')

const packages = [inefficientPackage]
const packagesUsingIds = packages.map((p) => {
  return {
    packaging: p.packaging._id,
    content: p.content._id
  }
})

const ids: {_id: ObjectId}[] = await orm.em.aggregate(entities.Order, [{ $match: { packages: packagesUsingIds }}, { $project: {_id: 1} }])
const foundOrder = await orm.em.findOne(entities.Order, {_id: ids[0]._id})
console.log(foundOrder)
```

**Reply by B4nan** [reply]
You can do something like `em.getCollect(Order).find()` and map the result to entities via `em.map()`.

**Reply by B4nan** [reply]
Actually, the `$eq: []` would make a lot of sense here, we can easily catch that and disable the `$in` operator expansion as nesting `$eq` with `$in` clearly does not make sense.

Let me know if latest dev version works: https://github.com/mikro-orm/mikro-orm/commit/e25d28e7c12a655d2373e14fd74e9cc68c2b1c5d

**Reply by edorgeville** [reply]
Seems to work! Except for the Typescript error:
```typescript
  console.log('find the order from the package')
  const packages = [inefficientPackage]
  console.log(await orm.em.findOne(entities.Order, { packages: { $eq: packages } })) // null

  console.log('find the order from the package ids')
  const packagesUsingIds = packages.map((p) => {
    return {
      packaging: p.packaging._id,
      content: p.content._id
    }
  })
  // @ts-expect-error Argument of type '{ packages: { $eq: { packaging: ObjectId; content: ObjectId; }[]; }; }' is not assignable to parameter of type 'FilterQuery<Order>'.
  console.log(await orm.em.findOne(entities.Order, { packages: { $eq: packagesUsingIds } })) // found!
```
Output:
```
find the order from the package
null
find the order from the package ids
Order {
  address: '10 Hello St.',
  packages: [ Package { packaging: [Packaging], content: [Product] } ],
  _id: ObjectId('6331a60b5130b5f0fc00a986')
}
```

**Reply by B4nan** [reply]
Weird, the `$eq` operator should now accept an array too at the type level: 

https://github.com/mikro-orm/mikro-orm/commit/e25d28e7c12a655d2373e14fd74e9cc68c2b1c5d#diff-26802e8761e5d043ee2361ebc1197aeff3393e3dfd04db6c070392337a4e9afdR61

How does the full error look like? Maybe we need to add `Partial` somewhere.

**Reply by edorgeville** [reply]
```
Argument of type '{ packages: { $eq: { packaging: ObjectId; content: ObjectId; }[]; }; }' is not assignable to parameter of type 'FilterQuery<Order>'.
  Types of property 'packages' are incompatible.
    Type '{ $eq: { packaging: ObjectId; content: ObjectId; }[]; }' is not assignable to type 'Package | Package[] | OperatorMap<Package> | ObjectQuery<Package> | FilterQuery<Package>[] | Package[]'.
      Types of property '$eq' are incompatible.
        Type '{ packaging: ObjectId; content: ObjectId; }[]' is not assignable to type 'Package | Package[]'.
          Type '{ packaging: ObjectId; content: ObjectId; }[]' is not assignable to type 'Package[]'.
            Type '{ packaging: ObjectId; content: ObjectId; }' is not assignable to type 'Package'.
              Types of property 'packaging' are incompatible.
                Type 'ObjectId' is missing the following properties from type 'Packaging': _id, width, height, format
```

**Reply by edorgeville** [reply]
Should I not be using ObjectIds?

**Reply by edorgeville** [reply]
Full repro is still here and up-to-date if you need to poke around
https://github.com/edorgeville/mikro-orm-find-embeddables

**Reply by B4nan** [reply]
The problem is that `$eq` expects a scalar value, or now possibly an array, but still of scalar values. Will see if I can make it work out of box for this use case too.
